CHOLMOD: Use CUDA as linker language if applicable

mmuetzel opened the pull request:
Use CUDA as the linker language if applicable while still avoiding Fortran as the linker language (see #799).

Addresses #929.

I seem to recall that there were CI runners that tested with CUDA using MSVC on Windows. It looks like there are none any more.
Also add one runner that tests building with CUDA on Windows (again?).

mmuetzel committed:
CI: Add runner with MSVC using CUDA.

mmuetzel committed:
CHOLMOD: Use CUDA as linker language if applicable.

Use CUDA as the linker language if applicable while still avoiding
Fortran as the linker language (see #799).

Addresses #929.

DrTimothyAldenDavis commented:
The cuda CI on MSVC was failing to load the cuda package-- a github or other bug, not mine-- so I had to disable it

mmuetzel commented:
> The cuda CI on MSVC was failing to load the cuda package-- a github or other bug, not mine-- so I had to disable it

I see.
I seem to recall that downloading the installer from the NVIDIA servers was slow sometimes for the GitHub runners. It still takes its time. But given its size (several Gigabyte), that is probably to be expected.

I'll try to think about what could be done to not fail the CI in case the NVIDIA servers are down (or slow) again.

In the meantime, is it ok to add a runner with CUDA on Windows again?

DrTimothyAldenDavis commented:
Yes, if it can work let's add it back in

mmuetzel commented:
I force pushed to trigger the CI again. 
If the installer is downloading also the second time, I'd say that it's worth enabling it for one runner again.

mmuetzel committed:
CHOLMOD: Resolve CUDA device symbols when linking static library.

If CUDA device symbols are not resolved when linking a static library
(with MSVC) dependent binaries need to be linked with the CUDA compiler
as the linker driver. That doesn't happen automatically with current
versions of CMake.

Instead, resolve CUDA device symbols when linking the static CHOLMOD
library. That means that no CUDA calls will be possible from dependent
binaries to CUDA functions in the static library.
As far as I can tell, no CUDA functions are part of the API of the
CHOLMOD library. So, doing that shouldn't be an issue.

mmuetzel committed:
CI (MSVC): Add runner that builds only static libraries with CUDA

mmuetzel commented:
The last two commits set `CUDA_RESOLVE_DEVICE_SYMBOLS` to `ON` for the static CHOLMOD library to work around an issue where some CUDA symbols aren't resolved correctly when linking that static library in dependent binaries with MSVC.
That has some side effects as explained in the commit message. But those side effects are probably not important for the CHOLMOD library.
I wonder if that flag should only be set when building with MSVC (or MSVC-compatible compilers).

The last of these two commits adds a runner configuration to the CI matrix for MSVC that should be affected by the issue from #929.

mmuetzel commented:
I had a look at the functions that are compiled to GPU code (in `cholmod_gpu_kernels.cu`). None of them are declared in the installed header.
I gather that means that they aren't expected to be callable directly from code that is linking to the built CHOLMOD library. (But only from within the CHOLMOD library itself.)

In that case, resolving the CUDA device symbols when linking the static CHOLMOD library shouldn't be an issue. So, the changes in this PR should be OK.
